### PR TITLE
Feat/d6 component

### PR DESCRIPTION
**Pull request by justindwyer6**
This PR adds a new component that provides an input element with all of the (surprisingly complex) logic necessary to limit its value from 1-6. Any character other than 1-6 will not be accepted. If you type a second character, it will just overwrite the first one, so you can easily update the number without needing to hit backspace.

I also replaced the sketchy one-off inputs that existed throughout the app with this new component, which helped clean up a few files.

### CODE REVIEW

File: src/components/BetweenTwoCastles/index.jsx [hunk 1]
```diff
@@ -1,55 +1,51 @@
...
+const BetweenTwoCastles = () => (
+  <>
+    <p>
+      Fill squares from the bottom up; #s on top must be lower than
+      those below. When you complete a column, gain a bonus. Score 1
+      star per completed row (each castle has 3 rows).
+    </p>
+    <div className="game">
+      {/* Row 1 */}
+      <b />
```
Comment (abrad45) [inline src/components/BetweenTwoCastles/index.jsx:17]: I know this is not the point, but what are all of these `<b />` tags for?

File: src/components/DSixInput/index.jsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import React, { useState } from "react";
+
+const DSixInput = ({ className }) => {
+  const [value, setValue] = useState("");
+
+  const sanitizeInput = (rawInput) => {
+    const currentValue = value;
+    const validValueWhitelist = ["", "1", "2", "3", "4", "5", "6"];
+
+    if (rawInput.length < 0 || rawInput.length > 2) {
```
Comment (abrad45) [inline src/components/DSixInput/index.jsx:10]: how could a string's length ever be `< 0`?

File: src/components/DSixInput/index.jsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      return currentValue;
+    }
+
+    const newValue = rawInput.replace(currentValue, "");
```
Comment (abrad45) [inline src/components/DSixInput/index.jsx:14]: I'm not entirely sure what this accounts for

**Review by abrad45 — APPROVED** [review]
You've done more testing of this than I have, but I'm not entirely sure about `sanitizeInput` and `preventInvalidKeyStrokes`

It feels like each could be replaced with some version of:

```js
// parse the user's keystroke as an int
// radix does not default to 10 (see link in PR)
const inputInt = parseInt(newInput, 10);

// 1. if it's not an int, ignore it
// 2. if it's 0 or below, ignore it
// 3. if it's above 6, ignore it
if (isNaN(inputInt) || inputInt < 1 || inputInt > 6) {
  return oldInput;
}

return newInput;
```

[parseInt on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt)

-----

with all this said, if you want to merge it, merge it :)

File: src/components/DSixInput/index.jsx (hunk 2 above)
Comment (justindwyer6) [reply src/components/DSixInput/index.jsx:10]: Very good question! 😅 Probably never.

File: src/components/DSixInput/index.jsx (hunk 3 above)
Comment (justindwyer6) [reply src/components/DSixInput/index.jsx:14]: `rawInput` is possibly a 2-digit number where one digit is equal to `currentValue`. This removes that digit and just leaves the newly typed one, that way you can update the number in the input without having to hit backspace.

File: src/components/BetweenTwoCastles/index.jsx (hunk 1 above)
Comment (justindwyer6) [reply src/components/BetweenTwoCastles/index.jsx:17]: I can't remember exactly why, but I think I read somewhere on the internet that it's a legit way to work with the need for an empty CSS Grid cell. It's been a while though, so not totally sure. I think it can make it a little easier to figure out which column an element will be placed in than specifying it in CSS rules.

Comment (abrad45) [reply src/components/BetweenTwoCastles/index.jsx:17]: Interesting. They're bold tags with no content. I wonder what about this tag makes it effective: my guess is that it's short and that a number of tags would work. Have you seen grid template areas? Super neat https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-areas

I'm nearly sure that these `b` tags are not accessible, but I don't think that needs to be a concern of this project.

**Comment by abrad45** [comment]
Summarizing our discussion on Discord, you said:

>  there's just no way around preventInvalidKeyStrokes that I can find. HTML seems to be weirdly stubborn about not letting React hijack those characters any other way. I tried using parseInt but React never even receives the value. Somehow HTML takes over when you enter one of those blacklist characters and React loses control of the input. It's super weird.

Maybe we should start with this: What are the requirements of DSixInput? If I understand correctly:

1. Allows you to enter a value from 1-6 (call this, `x`).
2. while the field already has a value:
    * Entering a new _valid_ (that is, `[1,6]`) value (call this, `y`) will replace `x` with `y`.
    * Entering a new _invalid_ value (anything else; again, `y`) will discard `y` and keep `x`.

Is there anything else? This should be the first step to determining what the best way to ensure our solution encompasses those needs

File: src/components/DSixInput/index.jsx (hunk 3 above)
Comment (justindwyer6) [reply src/components/DSixInput/index.jsx:8]: I'm not sure what you mean by "selecting the input on focus". I thought I explained the use case in my last comment. 🤔 I'm probably just misunderstand what you're asking about here. I'll answer your larger question in the other comment and maybe that will clarify things.

**Comment by justindwyer6** [comment]
> Summarizing our discussion on Discord, you said:
> 
> > there's just no way around preventInvalidKeyStrokes that I can find. HTML seems to be weirdly stubborn about not letting React hijack those characters any other way. I tried using parseInt but React never even receives the value. Somehow HTML takes over when you enter one of those blacklist characters and React loses control of the input. It's super weird.
> 
> Maybe we should start with this: What are the requirements of DSixInput? If I understand correctly:
> 
> 1. Allows you to enter a value from 1-6 (call this, `x`).
> 2. while the field already has a value:
>    
>    * Entering a new _valid_ (that is, `[1,6]`) value (call this, `y`) will replace `x` with `y`.
>    * Entering a new _invalid_ value (anything else; again, `y`) will discard `y` and keep `x`.
> 
> Is there anything else? This should be the first step to determining what the best way to ensure our solution encompasses those needs

@abrad45 Yep, those are essentially the requirements. Some other requirements are that:
1. An empty value should be allowed
2. Deleting the character should be allowed
3. Arrow keys to increment/decrement the value should be allowed
4. Pasting anything other than 1-6 should not be allowed

**Comment by abrad45** [comment]
It seems like the hardest requirement is that non-numerical input is ignored while values can be stepped through. This is a disagreement in the implementation of `input[type=number]`: See https://bugzilla.mozilla.org/show_bug.cgi?id=1398528

This codesandbox is as close as I can get in as succinct a way I can, but if you type `w`, nothing stops you. This seems to be specifically because of the following two props:

* `type="number"`: [[MDN]](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number) when this is turned off in my example code, typing `w` properly triggers the `onChange`, but you lose the stepper arrows. When it's turned on, the arrows are there, but typing `w` doesn't trigger `onChange` for reasons I could not begin to tell you. Interestingly, when it's turned off (with the `onChange` function I have, at least), typing a second numerical character after the first does nothing. I didn't look into why
* `inputMode="numeric"`: [[MDN]](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inputmode) doesn't seem to impact what the user can do on desktop Safari, but might on mobile. I did no browser testing of this code. 

https://codesandbox.io/s/vigorous-frog-wxg4i?file=/src/DSixInput.jsx

Two lines have `//@justin` at the end of them. You may uncomment those for some convenient logging.

---------

I raised this in the first place because my role at work where I am an engineering manager and not a developer is to look for code that is fragile or may be difficult to maintain and help everyone come to an agreement about what the best course of action is. Could we build this in a better way? Are all of these requirements truly requirements? I personally would prioritize relaxing a requirement or two to make the code simpler, but you are the manager of this project, not me, and realistically, we won't need to touch this stuff often.

*TLDR*: I have a better appreciation for the problem at hand and I'm comfortable with you merging your code as is. I do like the `useInput` pattern I've got in my codesandbox and encourage you to steal it or modify it however you like, but I don't want to hold up the merge here. Proceed as you wish. I support it either way :)
